api.lua

```diff
@@ -52,30 +83,23 @@ function areas:canInteractInArea(pos1, pos2, name, allow_open)
 		return true
 	end
 	areas:sortPos(pos1, pos2)
-	-- First check for a fully enclosing owned area.
-	if name then
-		for id, area in pairs(self.areas) do
-			-- A little optimization: isAreaOwner isn't necessary
-			-- here since we're iterating through all areas.
-			if area.owner == name and
-					self:isSubarea(pos1, pos2, id) then
-				return true
-			end
+
+	local areas = self:getAreasForArea(pos1, pos2)
+	for id, area in pairs(areas) do
+		-- First check for a fully enclosing owned area.
+		-- A little optimization: isAreaOwner isn't necessary
+		-- here since we're iterating through all relevant areas.
+		if area.owner == name and
+				self:isSubarea(pos1, pos2, id) then
+			return true
 		end
-	end
-	-- Then check for intersecting (non-owned) areas.
-	for id, area in pairs(self.areas) do
-		local p1, p2 = area.pos1, area.pos2
-		if (p1.x <= pos2.x and p2.x >= pos1.x) and
-		   (p1.y <= pos2.y and p2.y >= pos1.y) and
-		   (p1.z <= pos2.z and p2.z >= pos1.z) then
-			-- Found an intersecting area.
-			-- Return if the area is closed or open areas aren't
-			-- allowed, and the area isn't owned.
-			if (not allow_open or not area.open) and
-					(not name or not areas:isAreaOwner(id, name)) then
-				return false, id
-			end
+
+		-- Then check for intersecting (non-owned) areas.
+		-- Return if the area is closed or open areas aren't
+		-- allowed, and the area isn't owned.
+		if (not allow_open or not area.open) and
+				(not name or not self:isAreaOwner(id, name)) then
+			return false, id
```

Comment:
Combining these two loops into one like this won't work, as the function will return `false` if it finds a area that you don't own before finding a enclosing area that you do own.